File: agents/python/agent.py
"""
Basic Agent Framework for Messaging System

This module provides a simple framework for creating automated agents
that connect to the messaging system.
"""

import asyncio
import json
import random
import websockets
from datetime import datetime

class Agent:
    """
    Base Agent class for the messaging system.
    
    Concept: Agent Abstraction
    - Encapsulates the messaging protocol details
    - Provides simple methods for agent functionality
    - Handles connection and identity management
    """
    
    def __init__(self, broker_url="ws://localhost:8765", agent_name="GenericAgent"):
        """Initialize the agent with connection settings."""
        self.broker_url = broker_url
        self.agent_name = agent_name
        self.identity = None
        self.available_channels = []
        self.subscribed_channels = set()
    
    async def connect(self):
        """
        Connect to the message broker and receive identity.
        
        Concept: Connection Management
        - Establishes WebSocket connection
        - Processes identity assignment
        - Receives available channels
        """
        self.websocket = await websockets.connect(self.broker_url)
        
        # Receive and process identity
        identity_msg = await self.websocket.recv()
        if identity_msg.startswith("IDENTITY:"):
            self.identity = identity_msg.split(":")[1]
            print(f"Connected as: {self.identity}")
        else:
            raise Exception(f"Unexpected response: {identity_msg}")
        
        # Receive available channels
        channels_msg = await self.websocket.recv()
        if channels_msg.startswith("CHANNELS:"):
            channels = channels_msg.split(":")[1]
            self.available_channels = channels.split(",") if channels else []
            print(f"Available channels: {self.available_channels}")
    
    async def subscribe(self, channel):
        """
        Subscribe to a channel to receive messages.
        
        Concept: Channel Subscription
        - Agents join specific channels of interest
        - Enables topical message filtering
        """
        await self.websocket.send(f"SUBSCRIBE:{channel}")
        response = await self.websocket.recv()
        
        if response.startswith("SUB-ACK:"):
            self.subscribed_channels.add(channel)
            print(f"Subscribed to: {channel}")
            return True
        else:
            print(f"Failed to subscribe: {response}")
            return False
    
    async def publish(self, channel, message):
        """
        Publish a message to a channel.
        
        Concept: Message Distribution
        - Agents communicate by publishing to channels
        - Messages distributed to all subscribers
        """
        if channel not in self.subscribed_channels:
            await self.subscribe(channel)
            
        await self.websocket.send(f"PUBLISH:{channel}:{message}")
        
        # Wait for the message to be echoed back as confirmation
        while True:
            response = await self.websocket.recv()
            if response.startswith("MSG:"):
                _, msg_channel, sender, content = response.split(":", 3)
                if sender == self.identity and channel == msg_channel:
                    print(f"Message published to {channel}")
                    return True
    
    async def receive_messages(self, handler=None):
        """
        Process incoming messages, optionally with a custom handler.
        
        Concept: Event Handling
        - Agents can process and react to messages
        - Enables interactive behavior
        """
        async for message in self.websocket:
            print(f"Received: {message}")
            
            # Call custom handler if provided
            if handler and callable(handler):
                await handler(self, message)
    
    async def close(self):
        """Close the connection to the broker."""
        await self.websocket.close()
        print(f"Agent {self.identity} disconnected")


class ReportBot(Agent):
    """
    Example agent that generates and publishes reports.
    
    Concept: Specialized Agents
    - Extends base agent with specific functionality
    - Implements domain-specific behavior
    """
    
    def __init__(self, report_channel="dailyreports", **kwargs):
        super().__init__(**kwargs)
        self.report_channel = report_channel
    
    def generate_sales_report(self):
        """Generate a mock sales report with random data."""
        departments = ["Electronics", "Clothing", "Food", "Books"]
        
        report = {
            "date": datetime.now().strftime("%Y-%m-%d"),
            "total_sales": random.randint(5000, 20000),
            "departments": {}
        }
        
        for dept in departments:
            report["departments"][dept] = {
                "sales": random.randint(500, 5000),
                "transactions": random.randint(10, 100)
            }
        
        return report
    
    def format_report(self, report_data):
        """Format the report data as a readable message."""
        report_message = f"""
📊 DAILY SALES REPORT: {report_data['date']}
💰 Total Sales: ${report_data['total_sales']}

Department Breakdown:
{'-' * 30}
"""
        
        # Add each department to the message
        for dept, data in report_data["departments"].items():
            report_message += f"• {dept}: ${data['sales']} ({data['transactions']} transactions)\n"
            
        report_message += f"\nGenerated by {self.agent_name} at {datetime.now().strftime('%H:%M:%S')}"
        
        return report_message
    
    async def post_report(self):
        """Generate and post a report to the designated channel."""
        # Connect if not already connected
        if not hasattr(self, 'websocket'):
            await self.connect()
        
        # Subscribe to report channel if needed
        if self.report_channel not in self.subscribed_channels:
            await self.subscribe(self.report_channel)
        
        # Generate and format the report
        report_data = self.generate_sales_report()
        report_text = self.format_report(report_data)
        
        # Publish the report
        await self.publish(self.report_channel, report_text)
        
        return report_data


# Example usage
async def main():
    """Example of using the ReportBot."""
    # Create and configure the agent
    bot = ReportBot(agent_name="SalesReportBot")
    
    try:
        # Connect to the messaging system
        await bot.connect()
        
        # Post a report
        await bot.post_report()
        
        # Wait briefly to ensure message is processed
        await asyncio.sleep(2)
        
    finally:
        # Clean up
        await bot.close()


if __name__ == "__main__":
    # Run the example
    asyncio.run(main())

File: agents/go/agent.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"net/url"
	"os"
	"os/signal"
	"strings"
	"time"

	"github.com/gorilla/websocket"
)

// Agent represents a client that connects to the messaging system
type Agent struct {
	// Connection details
	BrokerURL     string
	AgentName     string
	Identity      string
	Conn          *websocket.Conn
	Subscriptions map[string]bool
	Channels      []string
}

// SalesReport represents a daily sales report structure
type SalesReport struct {
	Date        string                      `json:"date"`
	TotalSales  int                         `json:"total_sales"`
	Departments map[string]DepartmentReport `json:"departments"`
}

// DepartmentReport contains sales data for a department
type DepartmentReport struct {
	Sales        int `json:"sales"`
	Transactions int `json:"transactions"`
}

// NewAgent creates a new agent with the given configuration
func NewAgent(brokerURL, agentName string) *Agent {
	return &Agent{
		BrokerURL:     brokerURL,
		AgentName:     agentName,
		Subscriptions: make(map[string]bool),
	}
}

// Connect establishes a connection to the message broker
func (a *Agent) Connect() error {
	// Parse the URL
	u, err := url.Parse(a.BrokerURL)
	if err != nil {
		return fmt.Errorf("invalid URL: %v", err)
	}

	// Create dialer
	c, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
	if err != nil {
		return fmt.Errorf("dial error: %v", err)
	}
	a.Conn = c

	// Process initial messages from the server (identity and channels)
	err = a.processInitialMessages()
	if err != nil {
		return err
	}

	log.Printf("Agent %s connected as %s", a.AgentName, a.Identity)
	return nil
}

// processInitialMessages handles the identity and channels messages on connection
func (a *Agent) processInitialMessages() error {
	// Receive identity message
	_, identityMsg, err := a.Conn.ReadMessage()
	if err != nil {
		return fmt.Errorf("error reading identity message: %v", err)
	}

	// Process identity
	identityParts := strings.SplitN(string(identityMsg), ":", 2)
	if len(identityParts) != 2 || identityParts[0] != "IDENTITY" {
		return fmt.Errorf("unexpected identity format: %s", identityMsg)
	}
	a.Identity = identityParts[1]
	log.Printf("Received identity: %s", a.Identity)

	// Receive channels message
	_, channelsMsg, err := a.Conn.ReadMessage()
	if err != nil {
		return fmt.Errorf("error reading channels message: %v", err)
	}

	// Process channels
	channelsParts := strings.SplitN(string(channelsMsg), ":", 2)
	if len(channelsParts) != 2 || channelsParts[0] != "CHANNELS" {
		return fmt.Errorf("unexpected channels format: %s", channelsMsg)
	}

	if channelsParts[1] != "" {
		a.Channels = strings.Split(channelsParts[1], ",")
		log.Printf("Available channels: %v", a.Channels)
	} else {
		log.Printf("No channels available")
	}

	return nil
}

// Subscribe makes the agent join a channel
func (a *Agent) Subscribe(channel string) error {
	// Send subscribe message
	err := a.Conn.WriteMessage(websocket.TextMessage, []byte("SUBSCRIBE:"+channel))
	if err != nil {
		return fmt.Errorf("error subscribing to channel: %v", err)
	}

	// Wait for subscription acknowledgment
	_, msg, err := a.Conn.ReadMessage()
	if err != nil {
		return fmt.Errorf("error reading subscription response: %v", err)
	}

	// Check if subscription was successful
	if strings.HasPrefix(string(msg), "SUB-ACK:") {
		a.Subscriptions[channel] = true
		log.Printf("Subscribed to channel: %s", channel)
		return nil
	}

	return fmt.Errorf("failed to subscribe: %s", msg)
}

// Publish sends a message to a channel
func (a *Agent) Publish(channel, message string) error {
	// Check if subscribed to the channel
	if !a.Subscriptions[channel] {
		err := a.Subscribe(channel)
		if err != nil {
			return fmt.Errorf("can't publish without subscription: %v", err)
		}
	}

	// Send publish message
	err := a.Conn.WriteMessage(websocket.TextMessage, []byte("PUBLISH:"+channel+":"+message))
	if err != nil {
		return fmt.Errorf("error publishing message: %v", err)
	}

	log.Printf("Published message to channel %s", channel)
	return nil
}

// Close disconnects the agent from the broker
func (a *Agent) Close() error {
	if a.Conn != nil {
		// Send close message
		err := a.Conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
		if err != nil {
			log.Printf("Error during close message: %v", err)
		}
		return a.Conn.Close()
	}
	return nil
}

// GenerateSalesReport creates a random sales report
func (a *Agent) GenerateSalesReport() SalesReport {
	// Initialize random source
	rand.Seed(time.Now().UnixNano())

	// Create departments
	departments := []string{"Electronics", "Clothing", "Food", "Books"}
	deptReports := make(map[string]DepartmentReport)

	for _, dept := range departments {
		deptReports[dept] = DepartmentReport{
			Sales:        rand.Intn(4500) + 500, // 500-5000
			Transactions: rand.Intn(90) + 10,    // 10-100
		}
	}

	// Create report
	report := SalesReport{
		Date:        time.Now().Format("2006-01-02"),
		TotalSales:  rand.Intn(15000) + 5000, // 5000-20000
		Departments: deptReports,
	}

	return report
}

// FormatReport converts a sales report into a readable message
func (a *Agent) FormatReport(report SalesReport) string {
	var sb strings.Builder

	sb.WriteString(fmt.Sprintf("\n📊 DAILY SALES REPORT: %s\n", report.Date))
	sb.WriteString(fmt.Sprintf("💰 Total Sales: $%d\n\n", report.TotalSales))
	sb.WriteString("Department Breakdown:\n")
	sb.WriteString("------------------------------\n")

	// Add each department
	for dept, data := range report.Departments {
		sb.WriteString(fmt.Sprintf("• %s: $%d (%d transactions)\n", 
			dept, data.Sales, data.Transactions))
	}

	sb.WriteString(fmt.Sprintf("\nGenerated by %s at %s", 
		a.AgentName, time.Now().Format("15:04:05")))

	return sb.String()
}

// PostReport generates and posts a sales report
func (a *Agent) PostReport(channel string) error {
	// Generate report
	report := a.GenerateSalesReport()
	
	// Format the report
	formattedReport := a.FormatReport(report)
	
	// Publish to channel
	return a.Publish(channel, formattedReport)
}

// ListenForMessages processes incoming messages until interrupted
func (a *Agent) ListenForMessages() {
	for {
		_, message, err := a.Conn.ReadMessage()
		if err != nil {
			log.Printf("Read error: %v", err)
			return
		}
		log.Printf("Received: %s", message)
	}
}

func main() {
	// Create a new agent
	agent := NewAgent("ws://localhost:8765", "GoReportBot")
	
	// Connect to broker
	err := agent.Connect()
	if err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	
	// Ensure we disconnect properly
	defer agent.Close()
	
	// Subscribe to reports channel
	reportChannel := "dailyreports"
	err = agent.Subscribe(reportChannel)
	if err != nil {
		log.Fatalf("Failed to subscribe: %v", err)
	}
	
	// Post a report
	err = agent.PostReport(reportChannel)
	if err != nil {
		log.Fatalf("Failed to post report: %v", err)
	}
	
	// Set up interrupt handler
	interrupt := make(chan os.Signal, 1)
	signal.Notify(interrupt, os.Interrupt)
	
	// Process messages until interrupted
	msgDone := make(chan struct{})
	go func() {
		agent.ListenForMessages()
		close(msgDone)
	}()
	
	// Wait for interrupt or message processing to end
	select {
	case <-interrupt:
		log.Println("Interrupt received, closing connection")
	case <-msgDone:
		log.Println("Message processing ended")
	}
	
	log.Println("Agent completed successfully")
}
